Added bindings to JavascriptEditorViewport

arthursw opened the pull request:
Added bindings to `JavascriptEditorViewport` to be able to:
- react to mouse moves and inputs, and keyboard inputs,
- set the viewport type and the view mode,
- project/unproject from screen to world.

Also added a hack to be able to simply draw a selection rectangle on the viewport canvas in `FJavascriptEditorViewportClient::Draw(FViewport* Viewport, FCanvas* Canvas)`, with `SelectionBeginPoint` and `SelectionEndPoint`. This should be improved, but it would probably mean exposing FCanvas and FCanvasBoxItem to Blueprint/Javascript.

arthursw committed:
Added bindings to JavascriptEditorViewport to react to mouse moves and inputs, keyboard inputs, to set the viewport type and the view mode, and to project/unproject from screen to world

nakosung commented on Source/JavascriptEditor/JavascriptEditorViewport.cpp:
```diff
@@ -70,6 +124,19 @@ class FJavascriptEditorViewportClient : public FEditorViewportClient
+			FCanvasBoxItem Box(Widget->SelectionBeginPoint, Widget->SelectionEndPoint - Widget->SelectionBeginPoint);
+			Box.SetColor(FColor::White);
+			Box.LineThickness = 1.0;
+			Box.Draw(Canvas);
```
This function seems to handle draw behavior for specific use case. I think there should be some functions which exposes FCanvas*Item accessible to BP.

arthursw replied:
Yes, I fully agree, I updated my pull request.

I exposed the `FCanvas` as a `UCanvas` like in `JavascriptEdMode`.

For this purpose, I moved `GetCanvasByName()` from `JavascriptEdMode.cpp` to `JavascriptEditorLibrary.h` to be able to access it from `JavascriptEdMode.cpp` and `JavascriptEditorViewport.cpp`.

Tell me if there is a better solution.

arthursw replied:
I forgot to make some changes... Working on it...

arthursw committed:
Exposed DrawCanvas to EditorViewport

arthursw committed:
Merge remote-tracking branch 'upstream/master'

arthursw committed:
Merge branch 'ucanvas'

arthursw committed:
reformat/align code.

arthursw committed:
Update JavascriptEditorViewport.h

arthursw committed:
renamed DrawHUD into DrawCanvas

arthursw committed:
added GetWidgetMode()

nakosung commented:
It is recommended to hold UObject references by FGCObject. Could you transform your code not to call AddToRoot?

arthursw committed:
Removed CanvasObject->AddToRoot(), fixed problem in UJavascriptEditorViewport::ProjectWorldToScreen/DeprojectSceneToWorld()

arthursw commented:
Ok, I did something, I'm not sure that it is a good solution though.

I added 

```
UProperty()
UCanvas *Canvas;
```

to `UJavascriptEditorViewport` (I moved `GetCanvasByName()` back to `JavascriptEdMode.cpp`).

Instead, I could override `FEditorViewportClient::AddReferencedObjects()` for the `FJavascriptEditorViewportClient` (in `JavascriptEditorViewport.cpp`), maybe it is a better solution?

In `JavascriptEditorViewport.cpp`:

```
    virtual void AddReferencedObjects(FReferenceCollector& Collector) override
    {
        FEditorViewportClient::AddReferencedObjects(Collector);
        Collector.AddReferencedObject(CanvasObject);     // Add a reference to the UCanvas object
    }
```

(In the same way, I could also add `Collector.AddReferencedObject(CanvasObject);` in `FJavascriptEdModeInstance::AddReferencedObjects()` in `JavascriptEdMode.cpp`, and remove `CanvasObject->AddToRoot()` in `GetCanvasByName()`)

I also fixed some bugs in `UJavascriptEditorViewport::DeprojectScreenToWorld()`and `UJavascriptEditorViewport::ProjectWorldToScreen()`, and removed `FIntRect ViewRect;`and `FViewMatrices ViewMatrices;` from `UJavascriptEditorViewport`.

nakosung commented:
You may create a private object which inherits from FGCObject and override AddReferencedObjects. In this way, GC ref management can be done in right way. :)

I think uproperty only for gc isn't a good way.

arthursw commented:
Ok, but why it is better to create a private object which inherits from FGCObject than to override `FEditorViewportClient::AddReferencedObjects()`? (FEditorViewportClient is a FGCObject)

arthursw committed:
create a private object which inherits from FGCObject to hold UCanvas object

arthursw commented:
Is it better now?

nakosung commented:
Thanks, it is looking good to me!
